### PR TITLE
[#26061][Go SDK] Return server error on stream close for artifact upload.

### LINKED ISSUE #26061: [Failing Test]:  :sdks:go:test:ulrValidatesRunner appears to be very flaky at master
### What happened?

Sample failure runs: https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/2996

Also failed on some in-flight PRs.

```
13:19:29 --- PASS: TestEmitParDoAfterGBK (4.50s)
13:19:29 PASS
13:19:29 ok  	github.com/apache/beam/sdks/v2/go/test/regression	124.686s
13:19:29 FAIL
13:19:29 
13:19:29 > Task :sdks:go:test:ulrValidatesRunner FAILED
13:19:29 
13:19:29 FAILURE: Build failed with an exception.
13:19:29 
13:19:29 * Where:
13:19:29 Build file '/home/jenkins/jenkins-slave/workspace/beam_PreCommit_GoPortable_Cron/src/sdks/go/test/build.gradle' line: 152
13:19:29 
13:19:29 * What went wrong:
13:19:29 Execution failed for task ':sdks:go:test:ulrValidatesRunner'.
13:19:29 > Process 'command 'sh'' finished with non-zero exit value 1
13:19:29 
13:19:29 * Try:
13:19:29 > Run with --stacktrace option to get the stack trace.
13:19:29 > Run with --info or --debug option to get more log output.
```

### Issue Failure

Failure: Test is flaky

### Issue Priority

Priority: 1 (unhealthy code / failing or flaky postcommit so we cannot be sure the product is healthy)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [X] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by lostluck**
Adds retries for the portable artifact path, and corrects the error handling.

Python artifact server is closing the stream `StageViaPortableApi Recv error: rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: PROTOCOL_ERROR` which implies an implementation error on either the Python side, or the Go side of the divide, but most likely in the HTTP2 protocol handling, which is lower level than beam tends to go.

The Go half was following the sequence correctly, but wasn't logging server side close errors correctly. This PR surfaces allows those errors to be properly surfaced. It further adds retrying by restarting the reverse retrieval protocol. This should be sufficient to deflake the Go tests against the Portable Python runner, as the flake only happens rarely. But this isn't a for sure thing.

Also adds a test for the happy path on uploads. So retries aren't done, nor is the legacy backup path checked.

Closes #26061 for now, but we can always re-open if the same flakes recur, along with the improved logging.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Looks like middle of the day is sufficient!

```
2023/04/01 20:11:10 unable to stage with PortableAPI: failed to stage file /tmp/worker-11-1680379865888009992
	caused by:
chunk send failed
	caused by:
EOF; falling back to legacy
    ptest.go:108: Failed to execute job: 	staging artifacts
        failed to stage /tmp/worker-11-1680379865888009992 in 3 attempts: failed to send chunks for /tmp/worker-11-1680379865888009992; close error: rpc error: code = Unimplemented desc = Method not found!
        	caused by:
        EOF; failed to send chunks for /tmp/worker-11-1680379865888009992; close error: rpc error: code = Unimplemented desc = Method not found!
        	caused by:
        EOF; failed to send chunks for /tmp/worker-11-1680379865888009992; close error: rpc error: code = Unimplemented desc = Method not found!
        	caused by:
        EOF; failed to send chunks for /tmp/worker-11-1680379865888009992; close error: rpc error: code = Unimplemented desc = Method not found!
        	caused by:
        EOF
```

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
```
StageViaPortableApi Recv error: rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: PROTOCOL_ERROR
```

Which is apparently something that happens some times, but no clear resolution other than "the code is wrong some how", WRT the HTTP/2 path. Likely due to version skews between the Go and Python GRPC impleemntations? 

Monday I'll refactor this to add retries around the portable API path, which should resolve the flake, but keeping some of the extra logging.

**Comment by lostluck** [comment]
cc: @tvalentyn

File: sdks/go/pkg/beam/runners/universal/runnerlib/stage.go [hunk 1]
```diff
@@ -107,10 +139,10 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 	}
 }
 
-func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtifactRetrievalServiceClient) error {
+func stageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtifactRetrievalServiceClient) error {
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/runners/universal/runnerlib/stage.go:142]: I'm guessing we aren't expecting anyone to have been depending on this function being exported?

Comment (lostluck) [reply sdks/go/pkg/beam/runners/universal/runnerlib/stage.go:142]: Not in the least.

1. It's deep within the runner lib.
2. We're the only ones that depend on it in open source land.
3. It's using the fairly esoteric (WRT beam knowledge) Reverse Artifact Retreival service, which while it works, isn't something many have experience with.
4. It was brittle to begin with (erroring in ways that hid server connection closes).
5. It shouldn't have been exposed in the first place.
6. It was also undocumented, which doesn't encourage use...

**Pull request merged**